### PR TITLE
✨  Add 3DS outscale provider

### PR DESCRIPTION
**Pull request by outscale-vbr**
**What this PR does / why we need it**:
Add 3ds outscale provider
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: cmd/clusterctl/client/config/providers_client.go [hunk 1]
```diff
@@ -372,3 +380,37 @@ func validateProvider(r Provider) error {
 	}
 	return nil
 }
+
+func GetLastVersion(projectOwner string, projectName string) (string ) {
```
Comment (killianmuldoon) [inline cmd/clusterctl/client/config/providers_client.go:384]: Is there something different about outscale that it requires a different flow for getting the latest tag? Otherwise it's probably better to add the provider in the same way as the others and discuss / implement the change in versioning across all providers instead of just this one.

**Review by killianmuldoon — COMMENTED** [review]
/ok-to-test

File: cmd/clusterctl/client/config/providers_client.go [hunk 2]
```diff
@@ -111,7 +115,7 @@ func (p *providersClient) defaults() []Provider {
 	// out-of-box experience. This is an opt-in feature; other providers can be added by using the clusterctl configuration file.
 
 	// if you are a developer of a SIG-cluster-lifecycle project, you can send a PR to extend the following list.
-
+        outscaleVersion := GetLastVersion("outscale-dev", "cluster-api-provider-outscale")
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/config/providers_client.go:118]: all the other providers use latest in their github url and it automatically resolves to the latest released version. I strongly recommend using the same approach also here

**Review by ykakarap — COMMENTED** [review]
Here is an example for adding a new provider: https://github.com/kubernetes-sigs/cluster-api/pull/7001

File: docs/book/src/clusterctl/provider-contract.md [hunk 3]
```diff
@@ -257,7 +257,7 @@ providers.
 |CAPVCD        | cluster.x-k8s.io/provider=infrastructure-vcd           |
 |CAPX          | cluster.x-k8s.io/provider=infrastructure-nutanix       |
 |CAPZ          | cluster.x-k8s.io/provider=infrastructure-azure         |
-
+|CAPO          | cluster.x-k8s.io/provider=infrastructure-outscale      |
```
Comment (oscr) [inline docs/book/src/clusterctl/provider-contract.md:260]: CAPO is already used by OpenStack.

**Comment by outscale-vbr** [comment]
@oscr @fabriziopandini @ykakarap 
I change with latest but i have:
with /root/.cluster-api/clusterctl.yaml:
```
providers:
- name: outscale
  type: InfrastructureProvider
  url: https://github.com/outscale-dev/cluster-api-provider-outscale/releases/latest/infrastructure-components.yaml 
```
clusterctl init --infrastructure outscale
```
Fetching providers
Error: failed to get provider components for the "outscale" provider: failed to get repository client for the InfrastructureProvider with name outscale: error creating the GitHub repository client: failed to get GitHub latest version: failed to find releases tagged with a valid semantic version number  
```

whereas when i specify release version:
with /root/.cluster-api/clusterctl.yaml:
```
providers:
- name: outscale
  type: InfrastructureProvider
  url: https://github.com/outscale-dev/cluster-api-provider-outscale/releases/v0.1.0/infrastructure-components.yaml 
```
I have:
```
Fetching providers
Skipping installing cert-manager as it is already installed
Installing Provider="cluster-api" Version="v1.2.2" TargetNamespace="capi-system"
Installing Provider="bootstrap-kubeadm" Version="v1.2.2" TargetNamespace="capi-kubeadm-bootstrap-system"
Installing Provider="control-plane-kubeadm" Version="v1.2.2" TargetNamespace="capi-kubeadm-control-plane-system"
Installing Provider="infrastructure-outscale" Version="v0.1.0" TargetNamespace="cluster-api-provider-outscale-system"

Your management cluster has been initialized successfully!

You can now create your first workload cluster by running the following:

  clusterctl generate cluster [name] --kubernetes-version [version] | kubectl apply -f -
```
Release:
https://github.com/outscale-dev/cluster-api-provider-outscale/releases/tag/v0.1.0

Do you have some idea why it can not find latest ?

**Comment by outscale-vbr** [comment]
@killianmuldoon @ykakarap @oscr Thanks you, the issue was in metadata.yaml.

File: cmd/clusterctl/cmd/config_repositories_test.go [hunk 4]
```diff
@@ -117,6 +116,7 @@ digitalocean        InfrastructureProvider   https://github.com/kubernetes-sigs/
 docker              InfrastructureProvider   https://github.com/kubernetes-sigs/cluster-api/releases/latest/                             infrastructure-components-development.yaml
 gcp                 InfrastructureProvider   https://github.com/kubernetes-sigs/cluster-api-provider-gcp/releases/latest/                infrastructure-components.yaml
 hetzner             InfrastructureProvider   https://github.com/syself/cluster-api-provider-hetzner/releases/latest/                     infrastructure-components.yaml
+outscale            InfrastructureProvider   https://github.com/outscale-dev/cluster-api-provider-outscale/releases/latest/              infrastructure-components.yaml
```
Comment (ykakarap) [inline cmd/clusterctl/cmd/config_repositories_test.go:119]: I think this need to placed in the lexicographical order for the tests to pass.

File: cmd/clusterctl/cmd/config_repositories_test.go [hunk 5]
```diff
@@ -202,6 +202,10 @@ var expectedOutputYaml = `- File: core_components.yaml
   Name: hetzner
   ProviderType: InfrastructureProvider
   URL: https://github.com/syself/cluster-api-provider-hetzner/releases/latest/
+- File: infrastructure-components.yaml
+  Name: outscale
```
Comment (ykakarap) [inline cmd/clusterctl/cmd/config_repositories_test.go:206]: This needs to be placed in the lexicographical order for the tests to pass.

File: docs/book/src/user/quick-start.md [hunk 6]
```diff
@@ -366,6 +366,17 @@ export DO_B64ENCODED_CREDENTIALS="$(echo -n "${DIGITALOCEAN_ACCESS_TOKEN}" | bas
 clusterctl init --infrastructure digitalocean
 ```
 
+{{#/tab }}
+
+{{#tab Outscale}}
```
Comment (sbueringer) [inline docs/book/src/user/quick-start.md:371]: Must be added to the tab list above as well

File: cmd/clusterctl/client/config/providers_client.go [hunk 7]
```diff
@@ -113,7 +113,6 @@ func (p *providersClient) defaults() []Provider {
 	// out-of-box experience. This is an opt-in feature; other providers can be added by using the clusterctl configuration file.
 
 	// if you are a developer of a SIG-cluster-lifecycle project, you can send a PR to extend the following list.
-
```
Comment (sbueringer) [inline cmd/clusterctl/client/config/providers_client.go:116]: Please rollback unrelated formatting changes like this

File: cmd/clusterctl/client/config/providers_client.go [hunk 8]
```diff
@@ -21,7 +21,6 @@ import (
 	"os"
 	"sort"
 	"strings"
-
```
Comment (sbueringer) [inline cmd/clusterctl/client/config/providers_client.go:24]: Please fix the formatting findings reported by golagnci-lint.

(probably most of them should be fixed after running `make lint-fix`)

Comment (outscale-vbr) [reply cmd/clusterctl/client/config/providers_client.go]: Thank you, it is done

File: cmd/clusterctl/client/config/providers_client.go (hunk 7 above)
Comment (outscale-vbr) [reply cmd/clusterctl/client/config/providers_client.go]: Thanks you, it is done

File: docs/book/src/user/quick-start.md (hunk 6 above)
Comment (outscale-vbr) [reply docs/book/src/user/quick-start.md:371]: Thanks you, it is done

File: cmd/clusterctl/client/config/providers_client.go (hunk 1 above)
Comment (outscale-vbr) [reply cmd/clusterctl/client/config/providers_client.go]: I remove it.

File: docs/book/src/clusterctl/provider-contract.md (hunk 3 above)
Comment (outscale-vbr) [reply docs/book/src/clusterctl/provider-contract.md]: Change to CAPOSC

**Comment by fabriziopandini** [comment]
if I'm not wrong there are commits not belonging to this PR...

**Comment by outscale-vbr** [comment]
@fabriziopandini  I rebase on main

File: docs/book/src/user/quick-start.md [hunk 9]
```diff
@@ -665,6 +676,30 @@ export DO_NODE_MACHINE_TYPE=s-2vcpu-2gb
 export DO_NODE_MACHINE_IMAGE==<your-capi-image-id>
 ```
 
+{{#/tab }}
+
+{{#tab Outscale}}
```
Comment (sbueringer) [inline docs/book/src/user/quick-start.md:681]: Same here

File: docs/book/src/user/quick-start.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply docs/book/src/user/quick-start.md]: @sbueringer Thanks you, i change it.

[the remark this answers is not in the document]
Comment (outscale-vbr) [reply docs/book/src/user/quick-start.md]: Thanks for the command make serve-book

**Comment by sbueringer** [comment]
Thx!

/lgtm
/assign @fabriziopandini

**Comment by fabriziopandini** [comment]
thanks for your first PR  @outscale-vbr, looking forward to seeing more of your contributions!
/lgtm
/approve

**Comment by sbueringer** [comment]
@outscale-vbr 
Feel free to open a PR against release-1.2 if you want to get the changes into the v1.2.4 release (~ v1.2.4 is released in ca a month)
